authz/providers/azure/rbac/checkaccessreqhelper.go

```diff
@@ -0,0 +1,338 @@
+	Id               string `json:"id"`
+	RoleDefinitionId string `json:"roleDefinitionId"`
+	PrincipalId      string `json:"principalId"`
+	PrincipalType    string `json:"principalType"`
+	Scope            string `json:"scope"`
+	Condition        string `json:"condition"`
+	ConditionVersion string `json:"conditionVersion"`
+	CanDelegate      bool   `json:"canDelegate"`
+}
+
+type AzureRoleAssignment struct {
+	DelegatedManagedIdentityResourceId string `json:"delegatedManagedIdentityResourceId"`
+	RoleAssignment
+}
+
+type Permission struct {
+	Actions          []string `json:"actions,omitempty"`
+	NoActions        []string `json:"noactions,omitempty"`
+	DataActions      []string `json:"dataactions,omitempty"`
+	NoDataActions    []string `json:"nodataactions,omitempty"`
+	Condition        string   `json:"condition"`
+	ConditionVersion string   `json:"conditionVersion"`
+}
+
+type Principal struct {
+	Id   string `json:"id"`
+	Type string `json:"type"`
+}
+
+type DenyAssignment struct {
+	Id          string `json:"id"`
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	Permission
+	Scope                   string `json:"scope"`
+	DoNotApplyToChildScopes bool   `json:"doNotApplyToChildScopes"`
+	Principals              []Principal
+	ExcludePrincipals       []Principal
+	Condition               string `json:"condition"`
+	ConditionVersion        string `json:"conditionVersion"`
+}
+
+type AzureDenyAssignment struct {
+	MetaData          map[string]interface{} `json:"metadata"`
+	IsSystemProtected string                 `json:"isSystemProtected"`
+	IsBuiltIn         bool                   `json:"isBuiltIn"`
+	DenyAssignment
+}
+
+type AuthorizationDecision struct {
+	Decision            string              `json:"accessDecision"`
+	ActionId            string              `json:"actionId"`
+	IsDataAction        bool                `json:"isDataAction"`
+	AzureRoleAssignment AzureRoleAssignment `json:"roleAssignment"`
+	AzureDenyAssignment AzureDenyAssignment `json:"denyAssignment"`
+	TimeToLiveInMs      int                 `json:"timeToLiveInMs"`
+}
+
+func getScope(resourceId string, attr *authzv1.ResourceAttributes) string {
+	if attr != nil && attr.Namespace != "" {
+		return path.Join(resourceId, namespaces, attr.Namespace)
+	}
+	return resourceId
+}
+
+func isValidUUID(u string) bool {
+	_, err := uuid.Parse(u)
+	return err == nil
+}
+
+func getValidSecurityGroups(groups []string) []string {
+	var finalGroups []string
+	for _, element := range groups {
+		if isValidUUID(element) {
+			finalGroups = append(finalGroups, element)
+		}
+	}
+	return finalGroups
+}
+
+func getActionName(verb string) string {
+	/* special verbs
+	use verb on podsecuritypolicies resources in the policy API group
+	bind and escalate verbs on roles and clusterroles resources in the rbac.authorization.k8s.io API group
+	impersonate verb on users, groups, and serviceaccounts in the core API group
+	userextras in the authentication.k8s.io API group
```

Comment:
How is `userextras` relevant here?

Reply:
As we have taken dependency on Extra filed from Authentication stack which is mandatory for validating user access on Azure, user is required to pass userextras in kubeconfig if he wants to use impersonate.

Reply:
Ok. `userextras` is not a verb just the extra info `oid`. I was confused because it was listed under the `special verbs` headline. May be move it to the same line as `impersonate` verb?

Reply:
Yes, userextras would be resource and will be handled by impersonate. It is not much clear from the comment. Will update the comment.